core/src/protocol.rs

```diff
@@ -0,0 +1,194 @@
+use crate::platform::{AsyncReadExt, AsyncWriteExt, TlsStream};
+use crate::tdx::{self, TcgEvent};
+use crate::{AsyncByteStream, AttestationEndpoint, AttestationResult, Policy, RatlsError};
+use dcap_qvl::QuoteCollateralV3;
+use rand::rngs::OsRng;
+use rand::RngCore;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+fn success_default_true() -> bool {
+    true
+}
+
+#[derive(Serialize)]
+struct AttestationRequest {
+    report_data: String,
+}
+
+#[derive(Deserialize)]
+struct AttestationResponse {
+    #[serde(default = "success_default_true")]
+    success: bool,
+    quote: Option<DstackQuote>,
+    error: Option<String>,
+    #[serde(default)]
+    collateral: Option<QuoteCollateralV3>,
+}
+
+#[derive(Deserialize)]
+struct DstackQuote {
+    quote: String,
+    #[serde(default)]
+    event_log: Value,
+}
+
+pub async fn verify_attestation_stream<S>(
+    stream: &mut TlsStream<S>,
+    server_cert: &[u8],
+    policy: &Policy,
+    endpoint: &AttestationEndpoint,
+) -> Result<AttestationResult, RatlsError>
+where
+    S: AsyncByteStream,
+{
+    let mut nonce = [0u8; 32];
+    OsRng.fill_bytes(&mut nonce);
+    let nonce_hex = hex::encode(nonce);
+
+    let body_json = serde_json::to_string(&AttestationRequest {
+        report_data: nonce_hex.clone(),
+    })
+    .map_err(|e| RatlsError::Io(e.to_string()))?;
+    let request = format!(
+        "POST {} HTTP/1.1\r\n\
+         Host: {}\r\n\
+         Content-Type: application/json\r\n\
+         Content-Length: {}\r\n\
+         Connection: {}\r\n\
+         \r\n\
+         {}",
+        endpoint.path,
+        endpoint.host,
+        body_json.len(),
+        if endpoint.use_keep_alive {
+            "keep-alive"
+        } else {
+            "close"
+        },
```

Comment:
I already mentioned this above, but using `close` doesn't make sense in the context of RATLS IMO. The whole point of doing RATLS is to send data after.